### PR TITLE
Implement subpixel morphological antialiasing, or SMAA.

### LINKED ISSUE #9819: SMAA (Sub-pixel morphological antialiasing) support
## What problem does this solve or what need does it fill?
Feature: Support SMAA (Sub-pixel morphological antialiasing). It's a pracital implementation of MLAA (Morphological antialiasing).
Use Case: As far as I know, It's the best antialiasing that based on full-screen postprocessing only. It has the following advantages:

1. No temporal blur. TAA is a quite big trade-off: sometimes it looks absolutely amazing, sometimes horribly blur. There is even a dedicate subreddit for complaining about TAA: [/r/fuckTAA](https://www.reddit.com/r/FuckTAA/).
2. Render pipeline independent: Unlike MSAA, SMAA doesn't need you to change anything on the render pipeline. It's truly just a post-processing effect. Forward or deferred rendering doesn't matter.
3. Anti-aliasing on texture: MSAA only anti-aliases the boundaries of triangles. SMAA anti-aliases the aliasing cause by low-res textures too. (Of course it's a trade-off as well. Sometimes you don't want that.)
4. Looks better than FXAA: well never mind, anything looks better than FXAA.

## What solution would you like?

Implement a SMAA post processing effect.

## What alternative(s) have you considered?

TAA, MSAA, FXAA. But as I stated, SMAA has unique advantages over them.

## Additional context

Official SMAA implementation: https://github.com/iryoku/smaa (MIT license)
Paper: http://www.iryoku.com/smaa/
3D Graphics Discussion: https://github.com/bevyengine/bevy/discussions/4630 (I'm not sure if I should just posted there, but some items in that post link to separate issuse so I assume it's fine to open a new issue for further discussion/details)

### PR DESCRIPTION
**Pull request by pcwalton**
This commit implements a large subset of [*subpixel morphological antialiasing*], better known as SMAA. SMAA is a 2011 antialiasing technique that detects jaggies in an aliased image and smooths them out. Despite its age, it's been a continual staple of games for over a decade. Four quality presets are available: *low*, *medium*, *high*, and *ultra*. I set the default to *high*, on account of modern GPUs being significantly faster than they were in 2011.

Like the already-implemented FXAA, SMAA works on an unaliased image. Unlike FXAA, it requires three passes: (1) edge detection; (2) blending weight calculation; (3) neighborhood blending. Each of the first two passes writes an intermediate texture for use by the next pass. The first pass also writes to a stencil buffer in order to dramatically reduce the number of pixels that the second pass has to examine. Also unlike FXAA, two built-in lookup textures are required; I bundle them into the library in compressed KTX2 format.

The [reference implementation of SMAA] is in HLSL, with abundant use of preprocessor macros to achieve GLSL compatibility. Unfortunately, the reference implementation predates WGSL by over a decade, so I had to translate the HLSL to WGSL manually. As much as was reasonably possible without sacrificing readability, I tried to translate line by line, preserving comments, both to aid reviewing and to allow patches to the HLSL to more easily apply to the WGSL. Most of SMAA's features are supported, but in the interests of making this patch somewhat less huge, I skipped a few of the more exotic ones:

* The temporal variant is currently unsupported. This is and has been used in shipping games, so supporting temporal SMAA would be useful follow-up work. It would, however, require some significant work on TAA to ensure compatibility, so I opted to skip it in this patch.

* Depth- and chroma-based edge detection are unimplemented; only luma is. Depth is lower-quality, but faster; chroma is higher-quality, but slower. Luma is the suggested default edge detection algorithm. (Note that depth-based edge detection wouldn't work on WebGL 2 anyway, because of the Naga bug whereby depth sampling is miscompiled in GLSL. This is the same bug that prevents depth of field from working on that platform.)

* Predicated thresholding is currently unsupported.

* My implementation is incompatible with SSAA and MSAA, unlike the original; MSAA must be turned off to use SMAA in Bevy. I believe this feature was rarely used in practice.

The `anti_aliasing` example has been updated to allow experimentation with and testing of the different SMAA quality presets. Along the way, I refactored the example's help text rendering code a bit to eliminate code repetition.

SMAA is fully supported on WebGL 2.

Fixes #9819.

[*subpixel morphological antialiasing*]: https://www.iryoku.com/smaa/

[reference implementation of SMAA]: https://github.com/iryoku/smaa

### Added

* Subpixel morphological antialiasing, or SMAA, is now available. To use it, add the `SmaaSettings` component to your `Camera`.

Screenshot 2024-05-18 134311

### CODE REVIEW

**Comment by pcwalton** [comment]
How do I fix the typo false positives again?

**Comment by alice-i-cecile** [comment]
> How do I fix the typo false positives again? @mockersf

https://github.com/bevyengine/bevy/blob/main/typos.toml#L26

**Comment by pcwalton** [comment]
The CI failure is odd since it works here. I can try on Linux.

**Comment by pcwalton** [comment]
I can't reproduce the CI failure on Linux either. Help?

**Comment by pcwalton** [comment]
The problem is fixed and this is ready for review.

File: crates/bevy_core_pipeline/src/smaa/smaa.wgsl [hunk 1]
```diff
@@ -0,0 +1,1122 @@
...
+        weights = vec4(calculate_diag_weights(in.tex_coord, e, subsample_indices), weights.ba);
+
+        // We give priority to diagonals, so if we find a diagonal we skip 
+        // horizontal/vertical processing.
+        let diagonal_found = weights.r + weights.g != 0.0;
+#else   // SMAA_DISABLE_DIAG_DETECTION
+        let diagonal_found = false;
+#endif  // SMAA_DISABLE_DIAG_DETECTION
+
+        if (!diagonal_found) {
```
Comment (Elabajaba) [inline crates/bevy_core_pipeline/src/smaa/smaa.wgsl:988]: Imo doing an early return if a diagonal is found would be cleaner than this. It'd also let us get rid of the else statement on this function.

File: crates/bevy_core_pipeline/src/smaa/smaa.wgsl [hunk 2]
```diff
@@ -0,0 +1,1122 @@
...
+            // better interleave arithmetic and memory accesses):
+            d = abs(round(smaa_info.rt_metrics.zz * d - in.position.xx));
+
+            // SMAAArea below needs a sqrt, as the areas texture is compressed
+            // quadratically:
+            let sqrt_d = sqrt(d);
+
+            // Fetch the right crossing edges:
+            let e2 = textureSampleLevel(
+                edges_texture, edges_sampler, texture_offset(coords.zy, vec2<i32>(1, 0)), 0.0).r;
```
Comment (Elabajaba) [inline crates/bevy_core_pipeline/src/smaa/smaa.wgsl:1017]: Unnecessary, wgsl has an offset param for [`textureSampleLevel`](https://gpuweb.github.io/gpuweb/wgsl/#texturesamplelevel)

```suggestion
            let e2 = textureSampleLevel(edges_texture, edges_sampler, coords.zy, 0.0, vec2<i32>(1, 0)).r;
```

File: crates/bevy_core_pipeline/src/smaa/smaa.wgsl [hunk 3]
```diff
@@ -0,0 +1,1122 @@
...
+        // We want the distances to be in pixel units:
+        d = abs(round(smaa_info.rt_metrics.ww * d - in.position.yy));
+
+        // SMAAArea below needs a sqrt, as the areas texture is compressed
+        // quadratically:
+        let sqrt_d = sqrt(d);
+
+        // Fetch the bottom crossing edges:
+        let e2 = textureSampleLevel(
+            edges_texture, edges_sampler, texture_offset(coords.xz, vec2<i32>(0, 1)), 0.0).g;
```
Comment (Elabajaba) [inline crates/bevy_core_pipeline/src/smaa/smaa.wgsl:1060]:
```suggestion
        let e2 = textureSampleLevel(edges_texture, edges_sampler, coords.xz, 0.0, vec2<i32>(0, 1)).g;
```

Same as above.

File: crates/bevy_core_pipeline/src/smaa/smaa.wgsl [hunk 4]
```diff
@@ -0,0 +1,1122 @@
...
+ *    care of them.
+ * 2. The center pixel tap is reused.
+ *
+ * Note that in this case depth testing should be used instead of stenciling,
+ * as we have to write all the pixels in the motion blur pass.
+ *
+ * That's it!
+ */
+
+#import bevy_core_pipeline::fullscreen_vertex_shader::FullscreenVertexOutput
```
Comment (Elabajaba) [inline crates/bevy_core_pipeline/src/smaa/smaa.wgsl:295]: This is never used.

File: crates/bevy_core_pipeline/src/smaa/smaa.wgsl [hunk 5]
```diff
@@ -0,0 +1,1122 @@
...
+// of the clip-space rectangle that is at 1,-1 in clip space.
+fn calculate_vertex_varyings(vertex_index: u32) -> VertexVaryings {
+    // See the explanation above for how this works
+    let uv = vec2<f32>(f32(vertex_index >> 1u), f32(vertex_index & 1u)) * 2.0;
+    let clip_position = vec2<f32>(uv * vec2<f32>(2.0, -2.0) + vec2<f32>(-1.0, 1.0));
+
+    return VertexVaryings(clip_position, uv);
+}
+
+fn texture_offset(coords: vec2<f32>, offset: vec2<i32>) -> vec2<f32> {
```
Comment (Elabajaba) [inline crates/bevy_core_pipeline/src/smaa/smaa.wgsl:499]: This isn't needed because [`textureSampleLevel`](https://gpuweb.github.io/gpuweb/wgsl/#texturesamplelevel) has an offset parameter.

**Review by Elabajaba — COMMENTED** [review]
I haven't looked at the rust side code in smaa/mod.rs yet.

Originally this was filled with code changes for explicit `fma`s, but at least for AMD gpus the compiled GPU ISA was identical.

Some spots of the giant included readme don't make sense for the wgsl version (eg. section 6. mentioning SMAA_HLSL_*/SMAA_GLSL_*), but I'm not sure if that really matters.

File: crates/bevy_core_pipeline/src/smaa/smaa.wgsl [hunk 6]
```diff
@@ -0,0 +1,1122 @@
...
+        varyings.tex_coord.xyxy;
+    edge_detection_varyings.offset_2 = smaa_info.rt_metrics.xyxy * vec4(-2.0, 0.0, 0.0, -2.0) +
+        varyings.tex_coord.xyxy;
+
+    return edge_detection_varyings;
+}
+
+#endif  // SMAA_EDGE_DETECTION
+
+#ifdef SMAA_BLENDING_WEIGHT_CALCULATION
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_core_pipeline/src/smaa/smaa.wgsl:531]: They're necessary because otherwise there will be undefined references to uniforms that don't exist, since the per-stage uniforms are behind `#ifdef`s.

File: crates/bevy_core_pipeline/src/smaa/smaa.wgsl [hunk 7]
```diff
@@ -0,0 +1,1122 @@
...
+fn calculate_diag_weights(tex_coord: vec2<f32>, e: vec2<f32>, subsample_indices: vec4<f32>)
+        -> vec2<f32> {
+    var weights = vec2(0.0, 0.0);
+
+    // Search for the line ends:
+    var d = vec4(0.0);
+    var end = vec2(0.0);
+    if (e.r > 0.0) {
+        let d_xz = search_diag_1(tex_coord, vec2(-1.0, 1.0), &end);
+        d = vec4(d_xz.x, d.y, d_xz.y, d.w);
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_core_pipeline/src/smaa/smaa.wgsl:739]: Yeah, me too. It's easy to get wrong as well--I lost hours to debugging an issue from it!

File: crates/bevy_core_pipeline/src/smaa/smaa.wgsl [hunk 8]
```diff
@@ -0,0 +1,1122 @@
...
+//      -1  0  1  2  3
+//
+// The axes are clip-space x and y. The region marked s is the visible region.
+// The digits in the corners of the right-angled triangle are the vertex
+// indices.
+//
+// The top-left has UV 0,0, the bottom-left has 0,2, and the top-right has 2,0.
+// This means that the UV gets interpolated to 1,1 at the bottom-right corner
+// of the clip-space rectangle that is at 1,-1 in clip space.
+fn calculate_vertex_varyings(vertex_index: u32) -> VertexVaryings {
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_core_pipeline/src/smaa/smaa.wgsl:491]: That function is marked `@vertex` so I can't call it normally. (I tried.)

File: crates/bevy_core_pipeline/src/smaa/smaa.wgsl [hunk 9]
```diff
@@ -0,0 +1,1122 @@
...
+    // This function will unpack the values (mad + mul + round):
+    // wolframalpha.com: round(x * abs(5 * x - 5 * 0.75)) plot 0 to 1
+    var e = in_e;
+    e.r = e.r * abs(5.0 * e.r - 5.0 * 0.75);
+    return round(e);
+}
+
+fn decode_diag_bilinear_access_4(e: vec4<f32>) -> vec4<f32> {
+    let e_rb = e.rb * abs(5.0 * e.rb - 5.0 * 0.75);
+    return round(vec4(e_rb.x, e.g, e_rb.y, e.a));
```
[the remark this answers is not in the document]
Comment (pcwalton) [reply crates/bevy_core_pipeline/src/smaa/smaa.wgsl:669]: Yeah, I didn't like it either. If you have a better proposal I'd be happy to change it, but what I wrote there was the least bad thing I could think of.

File: crates/bevy_core_pipeline/src/smaa/smaa.wgsl (hunk 1 above)
Comment (pcwalton) [reply crates/bevy_core_pipeline/src/smaa/smaa.wgsl]: OK, I changed it to early return.

**Comment by pcwalton** [comment]
Addressed review comments.

**Review by Elabajaba — APPROVED** [review]

**Review by Olle-Lukowski — APPROVED** [review]
NOTE: This is the first bigger PR I am reviewing, so I would not count this as a full review, more like half a review.

I just spent a few hours reading, and trying to understand the code as well as possible. I think I understand the big picture, but not all of the finer details. But I tried the example, even tried it in some custom code, and the results are quite nice!

Code wise, I have no feedback other than good job, not too hard to read, proper comments are nice, etc.

TLDR: Good code quality, good results. I think this is ready to be merged, but I am not entitled to comment on implementation details.

**Comment by alice-i-cecile** [comment]
Thanks @Olle-Lukowski :) I'll take a closer look on Monday to check what I can myself.

**Comment by alice-i-cecile** [comment]
@pcwalton CI is complaining about the `typos.toml` file. I tried to resolve merge conflicts, but seem to have screwed up the formatting in some subtle way.

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1362 if you'd like to help out.
